examples/fluids/src/setupts.c

```diff
@@ -323,7 +339,11 @@ PetscErrorCode FormIJacobian_NS(TS ts, PetscReal t, Vec Q, Vec Q_dot,
     PetscCall(MatGetType(J_pre, &mat_type));
     if (strstr(mat_type, "kokkos") || strstr(mat_type, "cusparse"))
       mem_type = CEED_MEM_DEVICE;
-    CeedOperatorLinearAssemble(user->op_ijacobian, user->coo_values);
+    if (user->app_ctx->pmat_pbdiagonal) {
+      CeedOperatorLinearAssemblePointBlockDiagonal(user->op_ijacobian, user->coo_values, CEED_REQUEST_IMMEDIATE);
```

Comment:
@jeremylt @nbeams This fails on CPU for an "obvious" reason.
```console
$ ./navierstokes -options_file blasius.yaml -dm_plex_box_faces 40,60,1 -dm_plex_box_upper 4.2e-3,4.2e-3,5e-5 -{ts,snes}_converged_reason -snes_monitor -ts_max_steps 1 -output_freq 100 -ksp_converged_reason -pmat_pbdiagonal
[...]
/home/jed/src/libCEED/interface/ceed-elemrestriction.c:862 in CeedElemRestrictionApply(): Output vector size 125050 not compatible with element restriction (124050, 6000)
[...]
#0  0x00007ffff57f536c in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff57a5838 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff578f535 in abort () from /usr/lib/libc.so.6
#3  0x00007ffff5b9daca in CeedErrorAbort (ceed=<optimized out>, filename=<optimized out>, line_no=<optimized out>, func=<optimized out>, err_code=<optimized out>, format=0x7ffff5ccf558 "Output vector size %d not compatible with element restriction (%d, %d)", args=0x7fffffff9970) at /home/jed/src/libCEED/interface/ceed.c:1244
#4  0x00007ffff5b9e233 in CeedErrorImpl (ceed=<optimized out>, filename=filename@entry=0x7ffff5ccf358 "/home/jed/src/libCEED/interface/ceed-elemrestriction.c", lineno=lineno@entry=862, func=func@entry=0x7ffff5ccf740 <__func__.2> "CeedElemRestrictionApply", ecode=ecode@entry=2, format=format@entry=0x7ffff5ccf558 "Output vector size %d not compatible with element restriction (%d, %d)") at /home/jed/src/libCEED/interface/ceed.c:1167
#5  0x00007ffff5ba0460 in CeedElemRestrictionApply (rstr=<optimized out>, t_mode=t_mode@entry=CEED_TRANSPOSE, u=<optimized out>, ru=ru@entry=0x55555697bfb0, request=request@entry=0x7ffff5ea4828 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-elemrestriction.c:862
#6  0x00007ffff5baa09e in CeedSingleOperatorAssembleAddDiagonal (op=op@entry=0x5555569f5320, request=request@entry=0x7ffff5ea4828 <ceed_request_immediate>, is_pointblock=is_pointblock@entry=true, assembled=assembled@entry=0x55555697bfb0) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:500
#7  0x00007ffff5bab352 in CeedOperatorLinearAssembleAddPointBlockDiagonal (op=0x5555569f5320, assembled=0x55555697bfb0, request=0x7ffff5ea4828 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1768
#8  0x00007ffff5bab2ba in CeedOperatorLinearAssembleAddPointBlockDiagonal (op=0x555555e24f50, assembled=assembled@entry=0x55555697bfb0, request=request@entry=0x7ffff5ea4828 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1756
#9  0x00007ffff5bab329 in CeedCompositeOperatorLinearAssembleAddDiagonal (assembled=0x55555697bfb0, is_pointblock=true, request=0x7ffff5ea4828 <ceed_request_immediate>, op=0x1) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:539
#10 CeedOperatorLinearAssembleAddPointBlockDiagonal (op=op@entry=0x555555de5de0, assembled=assembled@entry=0x55555697bfb0, request=request@entry=0x7ffff5ea4828 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1765
#11 0x00007ffff5bab793 in CeedOperatorLinearAssemblePointBlockDiagonal (op=0x555555de5de0, assembled=0x55555697bfb0, request=0x7ffff5ea4828 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1700
#12 0x000055555557f386 in FormIJacobian_NS (ts=<optimized out>, t=<optimized out>, Q=<optimized out>, Q_dot=<optimized out>, shift=<optimized out>, J=<optimized out>, J_pre=0x555555cfbab0, user_data=<optimized out>) at /home/jed/src/libCEED/examples/fluids/src/setupts.c:343
#13 0x00007ffff7039ca9 in TSComputeIJacobian (ts=ts@entry=0x55555612ab80, t=1.9999999999999999e-06, U=U@entry=0x555555cf8570, Udot=0x555555cf9390, shift=shift@entry=500000, A=A@entry=0x555555cfbab0, B=B@entry=0x555555cfbab0, imex=imex@entry=PETSC_FALSE) at /home/jed/petsc/src/ts/interface/ts.c:946
#14 0x00007ffff7010657 in SNESTSFormJacobian_Theta (snes=<optimized out>, x=0x555555cf8570, A=0x555555cfbab0, B=0x555555cfbab0, ts=0x55555612ab80) at /home/jed/petsc/src/ts/impls/implicit/theta/theta.c:996

Reply:
The GPU version might be a simple fix to copy indices in `CeedElemRestrictionGetOffsets_Cuda`, where it will currently return NULL when like
```
(gdb) p {CeedElemRestriction_Cuda}rstr->data
$8 = {
  module = 0x5555593771e0,
  StridedTranspose = 0x5555594121e0,
  StridedNoTranspose = 0x5555593832b0,
  OffsetTranspose = 0x555559337cc0,
  OffsetNoTranspose = 0x555558fb2a00,
  num_nodes = 5002,
  h_ind = 0x0,
  h_ind_allocated = 0x0,
  d_ind = 0x7fff6dc3c400,
  d_ind_allocated = 0x7fff6dc3c400,
  d_t_offsets = 0x7fff6dc54000,
  d_t_indices = 0x7fff6dc59000,
  d_l_vec_indices = 0x7fff6dc4f000
}
```

Reply:
Wait, are there two separate issues?  Even if the offsets are NULL, that's different from the element restriction size mismatch error, right?

Reply:
`CeedSingleOperatorAssembleAddDiagonal` contains a diagonal and point-block diagonal path since 90% of the logic is the same between the two. Perhaps a slightly clearer name would be helpful

Reply:
The GPU backends have dedicated code for point-block diagonal, but this `h_ind` versus `d_ind` prevented us from running it in this case. I don't know if it would work for us otherwise.

Reply:
And I've been juggling too much so my assessment may be off.

Reply:
The restriction size mismatch could be coming from here:
https://github.com/CEED/libCEED/blob/bbd567d35929e5f8c26c293c1aa217f3636e617a/interface/ceed-preconditioning.c#L242

Reply:
It looks like this occurred in a version of the branch without #998 applied. I made a change in #988 that might be related.

Reply:
This fixes the size mismatch issue (which now produces identical results with `-pc_type vpbjacobi` using this lean assembly mode and full assembly).
https://github.com/CEED/libCEED/pull/998